source/Mlos.Python/mlos/Spaces/HypergridAdapters/unit_tests/TestContinuousToPolynomialBasisHypergridAdapter.py

```diff
@@ -0,0 +1,210 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+import math
+from mlos.Spaces import SimpleHypergrid, CategoricalDimension, ContinuousDimension, DiscreteDimension, OrdinalDimension
+from mlos.OptimizerEvaluationTools.SyntheticFunctions.ThreeLevelQuadratic import ThreeLevelQuadratic
+from mlos.Spaces.HypergridAdapters import ContinuousToPolynomialBasisHypergridAdapter
+
+
+class TestContinuousToPolynomialBasisHypergridAdapter:
+
+    @classmethod
+    def setup_class(cls) -> None:
+
+        cls.simple_hypergrid = SimpleHypergrid(
+            name='simple_adaptee',
+            dimensions=[
+                CategoricalDimension(name='categorical_mixed_types', values=['red', True, False, 5]),
+                DiscreteDimension(name='one_to_ten', min=1, max=10),
+                ContinuousDimension(name='z_one', min=-1, max=2),
+                ContinuousDimension(name='z_two', min=-2, max=1),
+                ContinuousDimension(name='z_3', min=-2, max=-1),
+                OrdinalDimension(name='ordinal_mixed_types', ordered_values=[1, False, 'two'])
+            ]
+        )
+
+        cls.unbalanced_hierarchical_hypergrid = SimpleHypergrid(
+            name='hierarchical_adaptee',
+            dimensions=[
+                CategoricalDimension(name='categorical_mixed_types', values=['red', True, False, 3]),
+                DiscreteDimension(name='one_to_ten', min=1, max=10),
+                ContinuousDimension(name='x1', min=-1, max=1),
+                ContinuousDimension(name='x2', min=-1, max=1),
+                OrdinalDimension(name='ordinal_mixed_types', ordered_values=[3, False, 'two'])
+            ]
+        ).join(
+            subgrid=SimpleHypergrid(
+                name="nested_grid",
+                dimensions=[
+                    CategoricalDimension(name='categorical_mixed_types', values=['red', False, True, 3]),
+                    DiscreteDimension(name='one_to_ten', min=1, max=10),
+                    ContinuousDimension(name='x1', min=-1, max=1),
+                    ContinuousDimension(name='x2', min=-1, max=1),
+                    OrdinalDimension(name='ordinal_mixed_types', ordered_values=[3, 'two', False])
+                ]
+            ),
+            on_external_dimension=CategoricalDimension("categorical_mixed_types", values=[True])
+        )
+
+        cls.balanced_hierarchical_hypergrid = ThreeLevelQuadratic().parameter_space
+
+    @staticmethod
+    def n_choose_k(n, k):
+        return math.factorial(n) / math.factorial(k) / math.factorial(n - k)
+
+    ## simple_hypergrid tests
+    # dataframe projections
+    def test_simple_hypergrid_degree_two_all_terms_dataframe(self):

```

Comment:
You might simplify some of these tests with ``pytest.mark.parametrize`` instead of making them different test functions, like
```python
@pytest.mark.parametrize("degree", [2, 3])
@pytest.mark.parametrize("interaction_only", [TrueFalse])
@pytest.mark.parmaetrize("adaptee", [self.simple_hypergrid, self.unbalanced_hierarchical_hypergrid, adaptee=self.balanced_hierarchical_hypergrid])
def test_dataframe_projection(degree, interaction_only, adaptee):
    # code of test_dataframe_projection
```

and similar for test_point_projection.

Reply:
Happy to use this strategy going forward.  @byte-sculptor do you agree?

Reply:
Looks quite neat :) I wouldn't block this PR on it, but down the road we should incorporate it. I think it can improve our test coverage.

Speaking of: @amueller what would it take to enable code coverage checks in GCI?

Reply:
#33 has some notes on that

Reply:
@amueller, @bpkroth , Am getting `NameError: name 'pytest' is not defined`.  How's this fixed?

Reply:
@byte-sculptor helped on this question, so please ignore my question. :)

Reply:
locally you can just set do ``pytest --cov-report html --cov=mlos``, the slightly trickier part is having the results hosted as a website, see the issue for a solution.

Reply:
I wouldn't necessarily block on it, but it makes the tests much shorter and easier to read in this instance and reduces the potential for mistakes by being much more consise.